_posts/2022-01-31-how-we-ship-ios-libraries.md

```diff
@@ -0,0 +1,333 @@
+```
+
+The `compile-library.sh` script is going to do the heavy lifting.
+The first argument is the crate name we want to compile, the second is the build variant to select.
+This is not yet defined, so let's do it first.
+
+Go to the "Build Settings" tab and click the `+` button to add a new "User-Defined Setting".
+Give it the name `buildvariant` and choose a value based on the build variant: `debug` for `Debug` and `release` for `Release`.
+
+Now we're ready to write the `compile-library.sh` script.
+Create a new directory `bin` in your project directory.
+In there create the file with the following content:
+
+```bash
+#!/usr/bin/env bash
+
+# This should be invoked from inside xcode, not manually
+if [ "$#" -ne 2 ]
+then
+    echo "Usage (note: only call inside xcode!):"
+    echo "Args: $*"
+    echo "path/to/build-scripts/xc-universal-binary.sh <FFI_TARGET> <buildvariant>"
+    exit 1
+fi
+
+# what to pass to cargo build -p, e.g. your_lib_ffi
+FFI_TARGET=$1
+# buildvariant from our xcconfigs
+BUILDVARIANT=$2
+
+RELFLAG=
+if [[ "$BUILDVARIANT" != "debug" ]]; then
+    RELFLAG=--release
+fi
+
+set -euvx
+
+if [[ -n "${DEVELOPER_SDK_DIR:-}" ]]; then
+  # Assume we're in Xcode, which means we're probably cross-compiling.
+  # In this case, we need to add an extra library search path for build scripts and proc-macros,
+  # which run on the host instead of the target.
+  # (macOS Big Sur does not have linkable libraries in /usr/lib/.)
+  export LIBRARY_PATH="${DEVELOPER_SDK_DIR}/MacOSX.sdk/usr/lib:${LIBRARY_PATH:-}"
+fi
+
+IS_SIMULATOR=0
+if [ "${LLVM_TARGET_TRIPLE_SUFFIX-}" = "-simulator" ]; then
+  IS_SIMULATOR=1
+fi
+
+for arch in $ARCHS; do
+  case "$arch" in
+    x86_64)
+      if [ $IS_SIMULATOR -eq 0 ]; then
+        echo "Building for x86_64, but not a simulator build. What's going on?" >&2
+        exit 2
+      fi
+
+      # Intel iOS simulator
+      export CFLAGS_x86_64_apple_ios="-target x86_64-apple-ios"
+      $HOME/.cargo/bin/cargo build -p $FFI_TARGET --lib $RELFLAG --target x86_64-apple-ios
+      ;;
+
+    arm64)
+      if [ $IS_SIMULATOR -eq 0 ]; then
+        # Hardware iOS targets
+        $HOME/.cargo/bin/cargo build -p $FFI_TARGET --lib $RELFLAG --target aarch64-apple-ios
+      else
+        $HOME/.cargo/bin/cargo build -p $FFI_TARGET --lib $RELFLAG --target aarch64-apple-ios-sim
+      fi
+  esac
+done
+```
+
+It's a bit long, but not too complex:
+It selects the cargo profile to use (that is whether to pass `--release` or not),
+sets up the `LIBRARY_PATH` if necessary and finally compiles the Rust library for the selected target.
+Xcode passes the architectures to build in `ARCHS`.
+It's either `x86_64` for simulator builds on Intel Mac hardware or `arm64`.
+If it's `arm64` it can be either the simulator or an actual hardware target.
+Those differ, but we can know which is which from what's in `LLVM_TARGET_TRIPLE_SUFFIX` and select the right Rust target.
```

Comment:
I'd be tempted to move this explanation in front of the script to front-load context before readers try to parse `bash` syntax (but that's me).